### PR TITLE
File loading

### PR DESCRIPTION
**Pull request by willrogers**
@TimGuiteDiamond two changes here:

* use React Context to allow any widget access to the base url if it wants it
* refactor `fromBob` and `fromJson` into one `embeddedDisplay` widget

### CODE REVIEW

File: public/json/synoptic.json [hunk 1]
```diff
@@ -2,15 +2,15 @@
   "type": "display",
   "position": "relative",
   "width": "600px",
-  "height": "200px",
+  "height": "210px",
```
Comment (TimGuiteDiamond) [inline public/json/synoptic.json]: Was this a very important 10 pixels? :sweat_smile:

File: src/app.tsx [hunk 2]
```diff
@@ -21,6 +22,8 @@ try {
   settings = {};
 }
 
+const baseUrl = settings.baseUrl ? settings.baseUrl : "http://localhost:3000";
```
Comment (TimGuiteDiamond) [inline src/app.tsx]: Is this a good place to use the "null coalescing" we now have access to (TS 3.7)?

`settings.baseUrl ?? "http://localhost:300"` ??

File: src/ui/widgets/EmbeddedDisplay/bobUtils.ts [hunk 3]
```diff
@@ -8,6 +8,24 @@ import { WidgetDescription } from "../createComponent";
...
+
+const OPI_WIDGET_MAPPING: { [key: string]: string } = {
+  "org.csstudio.opibuilder.Display": "display",
+  "org.csstudio.opibuilder.widgets.TextUpdate": "readback",
+  "org.csstudio.opibuilder.widgets.TextInput": "input",
+  "org.csstudio.opibuilder.widgets.Label": "label",
+  "org.csstudio.opibuilder.widgets.groupingContainer": "grouping",
+  "org.csstudio.opibuilder.widgets.Rectangle": "shape",
+  "org.csstudio.opibuilder.widgets.ActionButton": "actionbutton" // eslint-disable-line @typescript-eslint/camelcase
+};
```
Comment (TimGuiteDiamond) [inline src/ui/widgets/EmbeddedDisplay/bobUtils.ts]: This definitely makes sense :smiley:

File: src/ui/widgets/EmbeddedDisplay/bobUtils.ts [hunk 4]
```diff
@@ -151,9 +170,24 @@ export const bobChildToWidgetChild = (
     ...remainingProps
   } = bobChild;
 
+  let type = actuallyOpi ? _attributes.typeId : _attributes.type;
+  log.debug(`opi? ${actuallyOpi}`);
+  log.debug(`type is ${type}`);
+  log.debug(_attributes);
```
Comment (TimGuiteDiamond) [inline src/ui/widgets/EmbeddedDisplay/bobUtils.ts]: Can you add a comment about what this code is testing for?

File: public/json/synoptic.json (hunk 1 above)
Comment (willrogers) [reply public/json/synoptic.json]: Yup.

File: src/app.tsx (hunk 2 above)
Comment (willrogers) [reply src/app.tsx]: Yes, exactly.

File: public/json/synoptic.json (hunk 1 above)
Comment (TimGuiteDiamond) [reply public/json/synoptic.json]: :+1:

**Pull request merged**